src/Control/Carrier/NonDet/Church.hs

```diff
@@ -108,5 +110,6 @@ instance MonadTrans NonDetC where
 instance Algebra sig m => Algebra (NonDet :+: sig) (NonDetC m) where
   alg (L (L Empty))      = empty
   alg (L (R (Choose k))) = k True <|> k False
-  alg (R other)          = NonDetC $ \ fork leaf nil -> alg (handle (pure ()) (runNonDet (liftA2 (<|>)) runNonDetA (pure empty)) other) >>= runNonDet fork leaf nil
+  alg (R other)          = NonDetC $ \ fork leaf nil -> alg (handle (pure ()) dst other) >>= runIdentity . runNonDet (coerce fork) (coerce leaf) (coerce nil) where
+    dst = runIdentity . runNonDet (liftA2 (liftA2 (<|>))) (Identity . runNonDetA) (pure (pure empty))
```

Comment:
The old distributive law:

```haskell
dst :: forall a . NonDetC m (NonDetC m a) -> m (NonDetC m a)
dst = runNonDet (liftA2 (<|>)) runNonDetA (pure empty)
```

vs. the new one:

```haskell
dst :: forall a . NonDetC Identity (NonDetC m a) -> m (NonDetC Identity a)
dst = runIdentity . runNonDet (liftA2 (liftA2 (<|>))) (Identity . runNonDetA) (pure (pure empty))
```

The former is cute, but violates the (unenforced, at present) invariant that the context functor for a given algebra for a monad transformer `t` should be uniquely determined by `t`, i.e. it should not involve `m`. That’s sort of a moral invariant, in the sense that we’d only actually have to enforce it if we ever wanted to define a `MonadTransControl` instance for `NonDetC`, but we can see its effect in #296 where it forces `semantic` to list the entire monad stack underneath `NonDetC` in a `CanHandle` constraint.

IMO it also represents a better separation of concerns, and thus better modularity: `NonDetC Identity` explicitly tells us that everything about `NonDetC`’s effects can be captured in a context functor independent of whatever it wraps.

Reply:
Love this. Stick that `forall a . NonDetC Identity (NonDetC m a) -> m (NonDetC Identity a)` as an explicit type signature inside here and you’ve got a stew going.